netty/src/main/java/io/grpc/netty/ProtocolNegotiators.java

```diff
@@ -187,6 +190,55 @@ public AsciiString scheme() {
     }
   }
 
+
+  static final class ProxyNegotiator implements ProtocolNegotiator {
+    private final ProxyHandler proxyHandler;
+    private final ProtocolNegotiator negotiator;
+
+    public ProxyNegotiator(ProxyHandler proxyHandler, ProtocolNegotiator negotiator) {
+      this.negotiator = negotiator;
+      this.proxyHandler = proxyHandler;
+    }
+
+    @Override
+    public Handler newHandler(GrpcHttp2ConnectionHandler http2Handler) {
+      return new WithProxyHandler(proxyHandler, negotiator.newHandler(http2Handler));
+    }
+  }
+
+  static final class WithProxyHandler extends AbstractBufferingHandler
+      implements ProtocolNegotiator.Handler {
+    private ProxyHandler proxyHandler;
+
+    public WithProxyHandler(ProxyHandler proxyHandler, ChannelHandler... handlers) {
+      super(handlers);
+      checkNotNull(proxyHandler, proxyHandler);
```

Comment:
The second argument should probably be the string `"proxyHandler"`. We typically prefer doing it inline with the assignment to make it obvious that the variable is initialized non-null, but it's not that big of a deal:
```
this.proxyHandler = checkNotNull(proxyHandler, "proxyHandler");
```